lms/djangoapps/courseware/tests/test_anonymous.py

```diff
@@ -0,0 +1,78 @@
+# coding=utf-8
+"""
+Tests for public access to course content.
+"""
+from datetime import timedelta
+
+import ddt
+import mock
+from django.utils.timezone import now
+from xmodule.modulestore import ModuleStoreEnum
+from xmodule.modulestore.tests.django_utils import SharedModuleStoreTestCase
+from xmodule.modulestore.tests.factories import CourseFactory, ItemFactory
+
+from lms.djangoapps.courseware.url_helpers import get_redirect_url
+from openedx.core.djangoapps.waffle_utils import CourseWaffleFlag, WaffleFlagNamespace
+from openedx.core.djangoapps.waffle_utils.testutils import override_waffle_flag
+
+
+@ddt.ddt
+class AnonymousTestCase(SharedModuleStoreTestCase):
+    """
+    Testing public access to course content.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        """
+        Set up a course to be used for testing.
+        """
+        # setUpClassAndTestData() already calls setUpClass on SharedModuleStoreTestCase
+        # pylint: disable=super-method-not-called
+        with super(AnonymousTestCase, cls).setUpClassAndTestData():
+            with cls.store.default_store(ModuleStoreEnum.Type.split):
+                cls.course = CourseFactory.create(
+                    org='edX',
+                    number='test',
+                    display_name='Test Course',
+                    start=now() - timedelta(days=30),
+                )
+                with cls.store.bulk_operations(cls.course.id):
+                    chapter = ItemFactory.create(
+                        category='chapter',
+                        parent_location=cls.course.location,
+                        display_name='Test Chapter',
+                    )
+                    cls.section = ItemFactory.create(category='sequential', parent_location=chapter.location)
+                    section2 = ItemFactory.create(category='sequential', parent_location=chapter.location)
+                    ItemFactory.create(category='vertical', parent_location=cls.section.location)
+                    ItemFactory.create(category='vertical', parent_location=section2.location)
+
+    @ddt.data(
+        [False, 'private', 302],
+        [False, 'preview', 302],
+        [False, 'public', 302],
+        [True, 'private', 302],
+        [True, 'preview', 302],
+        [True, 'public', 200],
+    )
+    @ddt.unpack
+    def test_home_anonymous_user(self, waffle_override, course_visibility, expected_status_code):
```

Comment:
There is already a test for the flag: https://github.com/open-craft/edx-platform/blob/aa7599b49f1af7566f5a858c784f6b9687d439cd/lms/djangoapps/courseware/tests/test_views.py#L2332 . So instead of creating a new class, you can just add these ddt variations to it.

Also instead of patching the descriptor, you should be able to do `course = CourseFactory(course_visibility= course_visibility)`.

Reply:
Otherwise the test looks good.

Reply:
I've created a new class to perform a different setUp. It make sense to have all in a single place, but the code was getting very complex.

On the patching, do you mean creating a new course on each variation of the ddt?

Reply:
Can you explain where all the extra setup that is being done in this `setUpClass()` is being used? For the purposes of this test, what more do you need than what is already happening in https://github.com/open-craft/edx-platform/blob/aa7599b49f1af7566f5a858c784f6b9687d439cd/lms/djangoapps/courseware/tests/test_views.py#L2333-L2336 ?

Since the existing test is creating the course in the test, I meant when merging you can just add the parameter to the `CourseFactory()` call.

Reply:
Yeah, you recommendation made sense on a second look at the code, that's unnecessary.

I have updated the code, I'll just make the necessary changes to the other test and push the changes.